### PR TITLE
Added code reference validation and baseline function

### LINKED ISSUE #15: Use an AST to validate use statements
[nusje2000/phpstan-monolith](https://github.com/nusje2000/phpstan-monolith) uses phpstan to validate use statements in source code based on the composer.json, but this currently has a few limitations:
 - It is build with phpstan so the error's cannot be fixed automatically with an autofixer
 - PHPStan is significantly slower when using the extension
 - Errors can be confusing to read since one missing package can lead to hundreds of errors
 - Due to the caching that phpstan uses, updating the composer.json will not resolve the errors because the cached errors are shown

This should not be difficult to achieve using the nikic/PHP-Parser package.

### PR DESCRIPTION
**Pull request by nusje2000**
Fixes #15

### CODE REVIEW

File: src/Command/AbstractDependencyGraphCommand.php [hunk 1]
```diff
@@ -29,6 +31,16 @@ abstract class AbstractDependencyGraphCommand extends Command
...
     protected $io;
 
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    public function __construct()
+    {
+        parent::__construct();
```
Comment (dbrekelmans) [inline src/Command/AbstractDependencyGraphCommand.php:41]: Constructor only calling parent. Not very useful :)

File: src/Command/AbstractDependencyGraphCommand.php [hunk 2]
```diff
@@ -38,6 +50,7 @@ protected function configure(): void
     {
         $projectRoot = getcwd();
 
+        $this->logger = new ConsoleLogger($output);
```
Comment (dbrekelmans) [inline src/Command/AbstractDependencyGraphCommand.php:53]: I would put this as a constructor dependency to allow more flexibility in how the output is logged

File: src/Command/ValidateCommand.php [hunk 3]
```diff
@@ -44,30 +52,31 @@ protected function doExecute(DependencyGraph $graph): int
...
             return 0;
         }
 
-        foreach ($violations as $violation) {
+        foreach ($filteredViolations as $violation) {
             $this->io->writeln(sprintf('<violation>[VIOLATION]</violation> %s', $violation->getFormattedMessage()));
         }
 
-        if (!empty($violations)) {
+        if (!empty($filteredViolations)) {
```
Comment (aeviiq) [inline src/Command/ValidateCommand.php:67]: Unnecessary condition as you are already checking this on line 57 and returning if so.

File: src/Command/ValidateCommand.php [hunk 4]
```diff
@@ -82,8 +91,62 @@ protected function doExecute(DependencyGraph $graph): int
...
+        if (!file_exists($this->getBaselinePath($graph))) {
+            return $violations;
+        }
+
+        $contents = file_get_contents($this->getBaselinePath($graph));
+        if (false === $contents) {
+            throw new UnexpectedValueException('Could not read baseline.');
+        }
+
+        $ignored = Yaml::parse(file_get_contents($this->getBaselinePath($graph)) ?? '')['ignore'] ?? [];
```
Comment (aeviiq) [inline src/Command/ValidateCommand.php:115]: Unnecessary to retrieve the file contents again, as you already did on line 110.

File: src/Command/ValidateCommand.php [hunk 5]
```diff
@@ -82,8 +91,62 @@ protected function doExecute(DependencyGraph $graph): int
...
+        if ($input->getOption('static-analysis')) {
+            $rules->append(new Rule\OutOfBoundsClassReferenceRule($this->logger));
+        }
+
+        return $rules;
+    }
+
+    private function getBaselinePath(DependencyGraph $graph): string
+    {
+        return $graph->getRootPath() . DIRECTORY_SEPARATOR . 'composer-monolith-baseline.yaml';
```
Comment (aeviiq) [inline src/Command/ValidateCommand.php:150]: Would suggest to make a public constant for this file name, could be useful in the future to reference to it.

File: src/NodeVisitor/ClassUseVisitor.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        $this->packageFinder = $packageFinder;
+        $this->graph = $graph;
+        $this->violations = new ViolationCollection();
+    }
+
+    public function enterNode(Node $node): ?int
+    {
+        if (!$node instanceof Node\Name) {
+            return null;
```
Comment (aeviiq) [inline src/NodeVisitor/ClassUseVisitor.php:47]: Doesn't this imply incorrect usage of this method? If so, I would throw an exception.

File: src/Command/AbstractDependencyGraphCommand.php (hunk 1 above)
Comment (nusje2000) [reply src/Command/AbstractDependencyGraphCommand.php:41]: Ahh, I added that to add an instance of the logger but found out that it needed the input. Forgot to remove it.

**Comment by nusje2000** [comment]
TODO:
 - the code should be more optimized for large projects
 - ignore unreadable dirs
 - fix console output to work with input arguments to configure verbosity

**Comment by nusje2000** [comment]
This feature will be postponed to release in 2.1 since this still requires some work.

This means that release 2.0 will be released today.

**Pull request closed without merging**